### PR TITLE
Replace incremental data sinks with a single stream of TaskEvent enums

### PR DESCRIPTION
**Pull request by jspahrsummers**
~~_Depends on #22._~~

This is a refactoring that should've been done when we first moved from `ColdSignal` + `HotSignal` to `SignalProducer` + `Signal`.

The current sink-focused implementation bifurcates logic, and makes it really easy to introduce issues like that fixed by https://github.com/Carthage/ReactiveTask/pull/21. Handling everything within the same stream should make behavior more deterministic.

### CODE REVIEW

**Comment by robrix** [comment]
#22 has been merged.

**Comment by jspahrsummers** [comment]
:raised_hands:

File: ReactiveTask/Task.swift [hunk 1]
```diff
@@ -261,32 +385,76 @@ public func launchTask(taskDescription: TaskDescription, standardOutput: SinkOf<
...
-							if disposable.disposed {
-								stdoutPipe.closePipe()
-								stderrPipe.closePipe()
-
-								// Clean up the stdin pipe in a roundabout way.
-								stdinProducer.startWithSignal { signal, signalDisposable in
-									signalDisposable.dispose()
+							task.terminationHandler = { task in
+								let terminationStatus = task.terminationStatus
+								if terminationStatus == EXIT_SUCCESS {
```
Comment (norio-nomura) [inline ReactiveTask/Task.swift]: When `terminationStatus != EXIT_SUCCESS`, signal from `stderrProducer` does not wait signal from `stdoutProducer`. Signal from other hand does not be needed, but should be waited.
I guess this cause segmentation fault of https://github.com/Carthage/Carthage/issues/497

File: ReactiveTask/Task.swift [hunk 2]
```diff
@@ -261,32 +385,76 @@ public func launchTask(taskDescription: TaskDescription, standardOutput: SinkOf<
...
+									disposable += stdoutAggregated
+										|> map { data in .Success(Box(data)) }
+										|> start(observer)
+								} else {
+									disposable += stderrAggregated
+										|> flatMap(.Concat) { data in
+											let errorString = (data.length > 0 ? String(UTF8String: UnsafePointer<CChar>(data.bytes)) : nil)
+											return SignalProducer(error: .ShellTaskFailed(exitCode: terminationStatus, standardError: errorString))
+										}
+										|> start(observer)
```
Comment (norio-nomura) [inline ReactiveTask/Task.swift]: When `terminationStatus != EXIT_SUCCESS`, signal from `stderrProducer` does not wait signal from `stdoutProducer`. Signal from other hand does not be needed, but should be waited.
I guess this cause segmentation fault of https://github.com/Carthage/Carthage/issues/497

changed comment position

**Comment by jspahrsummers** [comment]
@norio-nomura Good catch yet again! I've pushed a fix for that issue.

**Comment by norio-nomura** [comment]
Let's go next https://github.com/Carthage/Carthage/pull/474/files#r31377774

**Comment by jspahrsummers** [comment]
There are still issues here, but I think it's in our interest to make a _less_ broken Carthage release available. I'd like to merge this, and continue to investigate the remaining issues separately.

Also, any help on that front would be greatly appreciated. :pray:

**Pull request merged**